### PR TITLE
Fix make_signature TypeError in py3

### LINKED ISSUE #17608: groupby isinstance(f, types.MethodType) in py3
in `core.groupby._whitelist_method_generator` (https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby.py#L2801)

```
    method_wrapper_template = \
        """def %(name)s(%(sig)s) :
    \"""
    %(doc)s
    \"""
    f = %(self)s.__getattr__('%(name)s')
    return f(%(args)s)"""
    property_wrapper_template = \
        """@property
def %(name)s(self) :
    \"""
    %(doc)s
    \"""
    return self.__getattr__('%(name)s')"""
    for name in whitelist:
        # don't override anything that was explicitly defined
        # in the base class
        if hasattr(GroupBy, name):
            continue
        # ugly, but we need the name string itself in the method.
        f = getattr(klass, name)
        doc = f.__doc__
        doc = doc if type(doc) == str else ''
        if isinstance(f, types.MethodType):
            wrapper_template = method_wrapper_template
            decl, args = make_signature(f)
            # pass args by name to f because otherwise
            # GroupBy._make_wrapper won't know whether
            # we passed in an axis parameter.
            args_by_name = ['{0}={0}'.format(arg) for arg in args[1:]]
            params = {'name': name,
                      'doc': doc,
                      'sig': ','.join(decl),
                      'self': args[0],
                      'args': ','.join(args_by_name)}
        else:
            wrapper_template = property_wrapper_template
            params = {'name': name, 'doc': doc}
        yield wrapper_template % params
```

It looks like `isinstance(f, types.MethodType)` is never evaluating True in py3, so these are all coming out as properties.

PY2:
```
In [1]: import pandas as pd
In [2]: gb = pd.Series(range(10)).groupby(lambda x: x % 2)
In [3]: gb.mad
Out[3]: <bound method SeriesGroupBy.mad of <pandas.core.groupby.SeriesGroupBy object at 0x10f394f10>>
In [4]: gb.mad.__doc__
Out[4]: '\n    \n\nReturn the mean absolute deviation of the values for the requested axis\n\nParameters\n----------\naxis : {index (0)}\nskipna : boolean, default True\n    Exclude NA/null values. If an entire row/column is NA, the result\n    will be NA\nlevel : int or level name, default None\n    If the axis is a MultiIndex (hierarchical), count along a\n    particular level, collapsing into a scalar\nnumeric_only : boolean, default None\n    Include only float, int, boolean columns. If None, will attempt to use\n    everything, then use only numeric data. Not implemented for Series.\n\nReturns\n-------\nmad : scalar or Series (if level specified)\n\n    '
```

PY3:
```
In [1]: import pandas as pd
In [2]: gb = pd.Series(range(10)).groupby(lambda x: x % 2)
In [3]: gb.mad
Out[3]: <function pandas.core.groupby._GroupBy._make_wrapper.<locals>.wrapper>
In [4]: gb.mad.__doc__
In [5]:
```

If we change the condition from `isinstance(f, types.MethodType)` to `not isinstance(f, property)`, a `TypeError` is raised (py3 only):

```
  File "[...]/pandas/core/groupby.py", line 2796, in _whitelist_method_generator
    decl, args = make_signature(f)
  File "[...]/pandas/util/_decorators.py", line 245, in make_signature
    defaults = ('',) * n_wo_defaults + spec.defaults
TypeError: can only concatenate tuple (not "list") to tuple
```

#17600 should make this problem irrelevant eventually, but someone might want to look at it more short-term.

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #17608 

- [ ] closes #xxxx
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/util/test_util.py [hunk 1]
```diff
@@ -467,3 +467,11 @@ def test_set_locale(self):
 
         current_locale = locale.getlocale()
         assert current_locale == self.current_locale
+
+
+def test_make_signature():
+    # See GH 17608
+    _ = make_signature(validate_kwargs)
```
Comment (jreback) [inline pandas/tests/util/test_util.py]: you need to assert something here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/util/test_util.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/util/test_util.py]: This test will fail under the status quo.

**Comment by jbrockmendel** [comment]
Travis error looks unrelated:

```
RuntimeError: module compiled against API version 0xb but this version of numpy is 0xa
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/pandas/lib/python3.6/site-packages/_pytest/config.py", line 342, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/home/travis/build/pandas-dev/pandas/pandas')
```

File: pandas/tests/util/test_util.py [hunk 2]
```diff
@@ -467,3 +467,11 @@ def test_set_locale(self):
...
         current_locale = locale.getlocale()
         assert current_locale == self.current_locale
+
+
+def test_make_signature():
+    # See GH 17608
+    make_signature(validate_kwargs)
+    # Case where the func does not have default kwargs
+    make_signature(deprecate_kwarg)
+    # Case where the func does have default kwargs
```
Comment (gfyoung) [inline pandas/tests/util/test_util.py]: Can you move your comments *above* the test you're running?

File: pandas/tests/util/test_util.py [hunk 3]
```diff
@@ -467,3 +467,11 @@ def test_set_locale(self):
 
         current_locale = locale.getlocale()
         assert current_locale == self.current_locale
+
+
+def test_make_signature():
+    # See GH 17608
+    # Case where the func does not have default kwargs
+    make_signature(validate_kwargs)
```
Comment (jreback) [inline pandas/tests/util/test_util.py]: what is this doing now and what does this fix?

Comment (jbrockmendel) [reply pandas/tests/util/test_util.py]: The case on line 476 raises a TypeError in py3.  See #17608.

Comment (jreback) [reply pandas/tests/util/test_util.py]: this works for me in py3 now, so its not a good enough example

Comment (jbrockmendel) [reply pandas/tests/util/test_util.py]: Before I dig into this, did you run `make_signature` with both cases in the test?  My last comment had a typo, should have said line 477.

Comment (jreback) [reply pandas/tests/util/test_util.py]: i ran make_signature(validate_kwargs)

Comment (jbrockmendel) [reply pandas/tests/util/test_util.py]: pls try the call in the next line `make_signature(deprecate_kwarg)`

Comment (jreback) [reply pandas/tests/util/test_util.py]: ok this is fine. but still put an expected result and test against that, otherwise lgtm.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green.

**Comment by jbrockmendel** [comment]
ping.  pls consider re-opening #17600, as it fixes a problem that caused this to fail silently.

**Comment by jreback** [comment]
thanks. see my comments for #17600

**Pull request merged**